contracts/CurrencyNetwork.sol

```diff
@@ -591,60 +706,113 @@ contract CurrencyNetwork is CurrencyNetworkInterface, Ownable, Authorizable, Des
+        _account.interestRateReceived = _interestRateReceived;
         _storeAccount(_creditor, _debtor, _account);
 
         emit TrustlineUpdate(
             _creditor,
             _debtor,
             _creditlineGiven,
-            _creditlineReceived
+            _creditlineReceived,
+            _interestRateGiven,
+            _interestRateReceived
         );
     }
 
     function _requestTrustlineUpdate(
         address _creditor,
         address _debtor,
-        uint32 _creditlineGiven,
-        uint32 _creditlineReceived
+        uint128 _creditlineGiven,
+        uint128 _creditlineReceived,
+        int16 _interestRateGiven,
+        int16 _interestRateReceived
     )
         internal
     {
         _storeTrustlineRequest(
             _creditor,
             _debtor,
-            TrustlineRequest(_creditlineGiven, _creditlineReceived, _creditor)
+            TrustlineRequest(_creditlineGiven, _creditlineReceived, _interestRateGiven, _interestRateReceived, _creditor)
         );
 
         emit TrustlineUpdateRequest(
             _creditor,
             _debtor,
             _creditlineGiven,
-            _creditlineReceived
+            _creditlineReceived,
+            _interestRateGiven,
+            _interestRateReceived
         );
     }
 
-    function _calculateFees(uint32 _value, int64 _balance, uint16 _capacityImbalanceFeeDivisor) internal pure returns (uint32) {
+    function _calculateFees(uint128 _value, int136 _balance, uint16 _capacityImbalanceFeeDivisor) internal pure returns (uint128) {
         if (_capacityImbalanceFeeDivisor == 0) {
             return 0;
         }
 
-        int64 imbalanceGenerated = int64(_value);
+        int136 imbalanceGenerated = int136(_value);
         if (_balance > 0) {
             imbalanceGenerated = _value - _balance;
             if (imbalanceGenerated <= 0) {
                 return 0;
             }
         }
-        return uint32(uint32(imbalanceGenerated / _capacityImbalanceFeeDivisor) + 1);  // minimum fee is 1
+        return uint128(uint128(imbalanceGenerated / _capacityImbalanceFeeDivisor) + 1);  // minimum fee is 1
+    }
+
+    function _calculateInterests(int136 _balance, uint32 _mtime, int16 _interestRateGiven, int16 _interestRateReceived) internal view returns (int136) {
+        int16 rate = defaultInterests;
+
+        if (customInterests) {
+            if (_balance > 0) {
+                rate = _interestRateGiven;
+            }else {
+                rate = _interestRateReceived;
+            }
+        }
+
+        return int136(_balance * int136(now - _mtime)/(60*60*24*365) * int136(rate)/100000);
+    }
+
+    // Calculates a representation of how much interests intermediaries (_b) gain participating in a transfer
+    // It is only calculating for one side and does not take into accounts the gains or loss from the whole mediation.
+    function _calculateInterestWeight(address _a, address _b, uint128 _transfer) internal view returns (int136) {
```

Comment:
The function is oblivious of this difference.
I call it with "forwardedValue" in _mediatedTransfer(), because it is the value impacting the balance so the interests generated by the creditline.